src/modules/ekf2/EKF/mag_control.cpp

```diff
@@ -41,199 +41,119 @@
 
-					_aid_src_mag_heading.time_last_fuse = _time_delayed_us;
-					_time_last_heading_fuse = _time_delayed_us;
-				}
-			}
-		}
+	if (_mag_bias_observable || _yaw_angle_observable) {
+		_time_last_mov_3d_mag_suitable = _time_delayed_us;
+	}
 
+	if (_params.mag_fusion_type == MagFuseType::NONE) {
+		stopMagFusion();
+		stopMagHdgFusion();
 		return;
 	}
 
-	if (_params.mag_fusion_type >= MagFuseType::NONE
-	    || _control_status.flags.mag_fault
-	    || !_control_status.flags.tilt_align) {
-
+	// stop mag (require a reset before using again) if there was an external yaw reset (yaw estimator, GPS yaw, etc)
+	if (_mag_decl_cov_reset && (_state_reset_status.reset_count.quat != _state_reset_count_prev.quat)) {
+		ECL_INFO("yaw reset, stopping mag fusion to force reinitialization");
 		stopMagFusion();
-		return;
+		resetMagCov();
 	}
 
-	_mag_yaw_reset_req |= !_control_status.flags.yaw_align;
+	magSample mag_sample;
 
-	if (mag_data_ready && !_control_status.flags.ev_yaw && !_control_status.flags.gps_yaw) {
+	if (_mag_buffer && _mag_buffer->pop_first_older_than(_time_delayed_us, &mag_sample)) {
 
-		if (shouldInhibitMag() || !mag_data_valid) {
-			if (uint32_t(_time_delayed_us - _mag_use_not_inhibit_us) > (uint32_t)5e6) {
-				// If magnetometer use has been inhibited continuously then stop the fusion
-				stopMagFusion();
-			}
+		if (mag_sample.reset || (_mag_counter == 0)) {
+			// sensor or calibration has changed, reset low pass filter (reset handled by controlMag3DFusion/controlMagHeadingFusion)
+			_control_status.flags.mag_fault = false;
 
-			return;
+			_state.mag_B.zero();
+			resetMagCov();
+
+			_mag_lpf.reset(mag_sample.mag);
+			_mag_counter = 1;
 
 		} else {
-			_mag_use_not_inhibit_us = _time_delayed_us;
+			_mag_lpf.update(mag_sample.mag);
+			_mag_counter++;
 		}
 
-		const bool mag_enabled_previously = _control_status.flags.mag_hdg || _control_status.flags.mag_3D;
+		const bool starting_conditions_passing = (_params.mag_fusion_type != MagFuseType::NONE)
+				&& checkMagField(mag_sample.mag)
+				&& (_mag_counter > 5) // wait until we have more than a few samples through the filter
+				&& (_state_reset_status.reset_count.quat == _state_reset_count_prev.quat) // no yaw reset this frame
+				&& (_control_status.flags.yaw_align == _control_status_prev.flags.yaw_align) // no yaw alignment change this frame
+				&& (_state_reset_status.reset_count.quat == _state_reset_count_prev.quat) // don't allow starting on same frame as yaw  reset
+				&& isNewestSampleRecent(_time_last_mag_buffer_push, MAG_MAX_INTERVAL);
 
-		// Determine if we should use simple magnetic heading fusion which works better when
-		// there are large external disturbances or the more accurate 3-axis fusion
-		switch (_params.mag_fusion_type) {
-		default:
-		// FALLTHROUGH
-		case MagFuseType::AUTO:
-			selectMagAuto();
-			break;
+		if (!_control_status.flags.tilt_align && !_control_status.flags.yaw_align && starting_conditions_passing) {
```

Comment:
I know that it was already there before, but why do we want to reset the yaw if tilt is not aligned? What's the benefit?
During leveling, we can just fuse a constant position and wait for it to complete before doing the yaw alignment.